### PR TITLE
Tagging support

### PR DESCRIPTION
**Pull request by bathina2**
## Change Overview

The following PR introduces the snapshot tagging for vsphere

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues

- #XXX

## Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/blockstorage/vmware/vmware.go [hunk 1]
```diff
@@ -358,14 +362,52 @@ func (p *FcdProvider) setTagsVolume(ctx context.Context, volume *blockstorage.Vo
...
 	return nil, errors.New("Not implemented")
 }
 
 // SnapshotsList is part of blockstorage.Provider
 func (p *FcdProvider) SnapshotsList(ctx context.Context, tags map[string]string) ([]*blockstorage.Snapshot, error) {
-	return nil, errors.New("Not implemented")
+	var snapshotIDs []types.ID
+	var err error
+	if len(tags) == 1 {
+		for k, v := range tags {
```
Comment (denisvmedia) [inline pkg/blockstorage/vmware/vmware.go:392]: These two lines look suspicious to me. If there is now error in the `if` condition, then probably there is no need in `for`.

File: pkg/blockstorage/vmware/vmware.go [hunk 2]
```diff
@@ -96,10 +101,20 @@ func NewProvider(config map[string]string) (blockstorage.Provider, error) {
 	if err != nil {
 		return nil, errors.Wrap(err, "Failed to create VSLM client")
 	}
+	c := rest.NewClient(cli)
+	err = c.Login(ctx, url.UserPassword(username, password))
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to login to VAPI rest client")
+	}
+	tm := vapitags.NewManager(c)
```
Comment (carlbraganza) [inline pkg/blockstorage/vmware/vmware.go:109]: Expose the actual tagging cli

File: pkg/blockstorage/vmware/vmware.go [hunk 3]
```diff
@@ -96,10 +101,20 @@ func NewProvider(config map[string]string) (blockstorage.Provider, error) {
...
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create tag manager")
+	}
 	gom := vslm.NewGlobalObjectManager(vslmCli)
 	return &FcdProvider{
-		Cns: cnsCli,
-		Gom: gom,
+		Cns:        cnsCli,
+		Gom:        gom,
+		TagManager: tm,
```
Comment (carlbraganza) [inline pkg/blockstorage/vmware/vmware.go:117]: Expose the tagging client and hide your interface.

File: pkg/blockstorage/vmware/vmware.go [hunk 4]
```diff
@@ -52,8 +55,10 @@ var (
 
 // FcdProvider provides blockstorage.Provider
 type FcdProvider struct {
-	Gom *vslm.GlobalObjectManager
-	Cns *cns.Client
+	Gom        *vslm.GlobalObjectManager
+	Cns        *cns.Client
+	TagManager tagManager
```
Comment (carlbraganza) [inline pkg/blockstorage/vmware/vmware.go:60]: Expose the tagging service client too - the `FcdProvider` type is used in CBT which could need it over time.

Hide your internal interface variable - it is really tacky anyway to expose a variable to an internal interface.

File: pkg/blockstorage/vmware/vmware.go [hunk 5]
```diff
@@ -358,14 +386,181 @@ func (p *FcdProvider) setTagsVolume(ctx context.Context, volume *blockstorage.Vo
 	return nil
 }
 
+// GetOrCreateCategory takes a category name and attempts to get or create the category
+// it returns the category ID.
+func (p *FcdProvider) GetOrCreateCategory(ctx context.Context, categoryName string) (string, error) {
+	cat, err := p.TagManager.GetCategory(ctx, categoryName)
```
Comment (carlbraganza) [inline pkg/blockstorage/vmware/vmware.go:392]: Why not use `p.GetCategoryID` here?

File: pkg/blockstorage/vmware/vmware.go [hunk 6]
```diff
@@ -358,14 +386,181 @@ func (p *FcdProvider) setTagsVolume(ctx context.Context, volume *blockstorage.Vo
...
+// the tags are of the form volid:snapid:tag:value
+// these tags are assigned to a predefined category that is initialized by the FcdProvider
+type snapshotTag struct {
+	volid  string
+	snapid string
+	key    string
+	value  string
+}
+
+func (t *snapshotTag) print() string {
```
Comment (carlbraganza) [inline pkg/blockstorage/vmware/vmware.go:428]: This signature should be `String() string`.

File: pkg/blockstorage/vmware/vmware.go [hunk 7]
```diff
@@ -358,14 +386,181 @@ func (p *FcdProvider) setTagsVolume(ctx context.Context, volume *blockstorage.Vo
...
+	if err != nil {
+		return "", errors.Wrap(err, "Failed to find category")
+	}
+	return cat.ID, nil
+}
+
+// snapshotTag is the struct that will be used to create vmware tags
+// the tags are of the form volid:snapid:tag:value
+// these tags are assigned to a predefined category that is initialized by the FcdProvider
+type snapshotTag struct {
```
Comment (carlbraganza) [inline pkg/blockstorage/vmware/vmware.go:421]: Should use a prefix of `vmWare` or `VMware`.

Why not expose the type?

File: pkg/blockstorage/vmware/vmware.go [hunk 8]
```diff
@@ -358,14 +386,181 @@ func (p *FcdProvider) setTagsVolume(ctx context.Context, volume *blockstorage.Vo
...
+func (p *FcdProvider) parseTag(tag string) (*snapshotTag, error) {
+	parts := strings.Split(tag, ":")
+	if len(parts) != 4 {
+		return nil, errors.Errorf("Malformed tag (%s)", tag)
+	}
+	return &snapshotTag{parts[0], parts[1], parts[2], parts[3]}, nil
+}
+
+// setTagsSnapshot sets tags for a snapshot
+func (p *FcdProvider) setTagsSnapshot(ctx context.Context, snapshot *blockstorage.Snapshot, tags map[string]string) error {
```
Comment (carlbraganza) [inline pkg/blockstorage/vmware/vmware.go:445]: `setSnapshotTags`

File: pkg/blockstorage/vmware/vmware.go [hunk 9]
```diff
@@ -358,14 +386,181 @@ func (p *FcdProvider) setTagsVolume(ctx context.Context, volume *blockstorage.Vo
...
+			Name:       tag.print(),
+		})
+		if err != nil && !strings.Contains(err.Error(), "ALREADY_EXISTS") {
+			return errors.Wrapf(err, "Failed to create tag (%s) for categoryID (%s) ", tag, p.categoryID)
+		}
+	}
+	return nil
+}
+
+func (p *FcdProvider) deleteTagsSnapshot(ctx context.Context, snapshot *blockstorage.Snapshot) error {
```
Comment (carlbraganza) [inline pkg/blockstorage/vmware/vmware.go:471]: `deleteSnapshotTags`

File: pkg/blockstorage/vmware/vmware.go [hunk 10]
```diff
@@ -358,14 +386,181 @@ func (p *FcdProvider) setTagsVolume(ctx context.Context, volume *blockstorage.Vo
...
+
+func (t *snapshotTag) print() string {
+	volid := strings.ReplaceAll(t.volid, ":", "-")
+	snapid := strings.ReplaceAll(t.snapid, ":", "-")
+	key := strings.ReplaceAll(t.key, ":", "-")
+	value := strings.ReplaceAll(t.value, ":", "-")
+	return fmt.Sprintf("%s:%s:%s:%s", volid, snapid, key, value)
+}
+
+func (p *FcdProvider) parseTag(tag string) (*snapshotTag, error) {
```
Comment (carlbraganza) [inline pkg/blockstorage/vmware/vmware.go:436]: `parseSnapshotTag`

**Review by carlbraganza — CHANGES_REQUESTED** [review]
See comments.

**Review by carlbraganza — APPROVED** [review]

**Pull request merged**